### PR TITLE
CO-3594: -Added a new subsection "My sponsorship" in "My Financials".

### PR DESCRIPTION
**Pull request by coup1234**

### CODE REVIEW

File: website_compassion/template/my_account_donations.xml [hunk 1]
```diff
@@ -261,5 +289,19 @@
                 <a href="https://compassion.ch/donations/" class="btn btn-primary" target="_blank">Make a donation</a>
             </div>
         </template>
+        <template id="my_donations_upgrade" name="upgrade">
+            <t t-set="selected">donation</t>
+            <t t-set="title">My financials</t>
+            <t t-call="website_compassion.my_account_layout">
+               <center> Comfirm upgrade to 50 CHF ? </center>
+<!--                      <div class="row">-->
```
Comment (JonathanGuerne) [inline website_compassion/template/my_account_donations.xml]: avoid leaving comments in published code

File: website_compassion/template/my_account_donations.xml [hunk 2]
```diff
@@ -261,5 +289,19 @@
...
+                 <center>
+            <a t-attf-href="/my/donations/update?res=accepted&amp;recurring_contract=#{sponsor}" class="btn btn-primary">Yes</a>
+            <a t-attf-href="/my/donations/update?res=refused&amp;recurring_contract=#{sponsor}" class="btn btn-primary">No</a>
+<!--               </div>-->
```
Comment (JonathanGuerne) [inline website_compassion/template/my_account_donations.xml]: comment not needed

File: website_compassion/template/my_account_donations.xml [hunk 3]
```diff
@@ -149,7 +150,34 @@
...
+                <t t-foreach="paid_sponsor" t-as="sponsor">
+                    <div class="row">
+                        <div class="col-6">
+                            <label><t t-esc="sponsor.child_id.preferred_name"/></label>
+                        </div>
+                        <div class="col-3">
+                            <label><t t-raw="int(sponsor.total_amount)"/> <t t-esc="sponsor.invoice_line_ids[0].currency_id.name"/></label>
+                        </div>
+                        <div class="col-3" t-if="int(sponsor.total_amount) &lt;= 42">
+                          <a t-attf-href="/my/donations/update?recurring_contract=#{sponsor.id}" class="btn btn-primary">Upgrade to 50 <t t-esc="sponsor.invoice_line_ids[0].currency_id.name"/></a>
```
Comment (JonathanGuerne) [inline website_compassion/template/my_account_donations.xml:174]: If selected sponsorship < 42 CHF the upgrade wouldn't be all the way to 50 (with current implementation). 
Maybe a bit confusing for the end user ?

File: website_compassion/controllers/my_account.py [hunk 4]
```diff
@@ -320,6 +347,9 @@ def my_donations(self, form_id=None, **kw):
             len(sponsorship.filtered(lambda s: s.type == "S"))
             for sponsorship in sponsorships_by_group
         ]
+        paid_sponsor = (partner.contracts_fully_managed
+                        + partner.contracts_correspondant) \
+            .filtered(lambda a: a.state not in ["cancelled", "terminated"])
```
Comment (JonathanGuerne) [inline website_compassion/controllers/my_account.py:354]: Same as above
`contracts_correspondant` refers to contracts where current partner doesn't pay. `contracts_paid` seems more appropriate

File: website_compassion/controllers/my_account.py [hunk 5]
```diff
@@ -269,6 +269,33 @@ def my_child(self, state="active", child_id=None, **kwargs):
             # No child is selected, we pick the first one by default
             return request.redirect(f"/my/children?child_id={children[0].id}")
 
+    @route("/my/donations/update", type="http", auth="user", website=True)
+    def my_donations_update(self, recurring_contract=None, res=None, **kw):
+        partner = request.env.user.partner_id
+        contract = (partner.contracts_fully_managed
+                    + partner.contracts_correspondant) \
```
Comment (JonathanGuerne) [inline website_compassion/controllers/my_account.py]: `contracts_correspondant` refers to contracts where current partner doesn't pay. `contracts_paid` seems more appropriate

File: website_compassion/template/my_account_donations.xml [hunk 6]
```diff
@@ -261,5 +289,19 @@
                 <a href="https://compassion.ch/donations/" class="btn btn-primary" target="_blank">Make a donation</a>
             </div>
         </template>
+        <template id="my_donations_upgrade" name="upgrade">
+            <t t-set="selected">donation</t>
+            <t t-set="title">My financials</t>
+            <t t-call="website_compassion.my_account_layout">
+               <center> Comfirm upgrade to 50 CHF ? </center>
```
Comment (JonathanGuerne) [inline website_compassion/template/my_account_donations.xml]: Same as above
If selected sponsorship < 42 CHF the upgrade wouldn't be all the way to 50 (with current implementation).
Maybe a bit confusing for the end user ?

File: website_compassion/controllers/my_account.py [hunk 7]
```diff
@@ -269,6 +269,33 @@ def my_child(self, state="active", child_id=None, **kwargs):
...
+            .filtered(lambda a: a.state not in ["cancelled", "terminated"]).filtered(
+            lambda a: int(a.id) == int(recurring_contract))
```
Comment (JonathanGuerne) [inline website_compassion/controllers/my_account.py:280]: Potential error when `recurring_contract` is not specified. Add redirection if None ?
a.id is already an int

File: website_compassion/controllers/my_account.py [hunk 8]
```diff
@@ -269,6 +269,35 @@ def my_child(self, state="active", child_id=None, **kwargs):
             # No child is selected, we pick the first one by default
             return request.redirect(f"/my/children?child_id={children[0].id}")
 
+    @route("/my/donations/update", type="http", auth="user", website=True)
```
Comment (ecino) [inline website_compassion/controllers/my_account.py:272]:
```suggestion
    @route("/my/donations/upgrade", type="http", auth="user", website=True)
```

File: website_compassion/controllers/my_account.py [hunk 9]
```diff
@@ -269,6 +269,35 @@ def my_child(self, state="active", child_id=None, **kwargs):
...
+    def my_donations_update(self, recurring_contract=None, res=None, **kw):
+        partner = request.env.user.partner_id
+        if not recurring_contract:
+            return request.redirect("/my/donations")
+        contract = (partner.contracts_fully_managed
+                    + partner.contracts_paid) \
+            .filtered(lambda a: a.state not in ["cancelled", "terminated"]).filtered(
+            lambda a: int(a.id) == int(recurring_contract))
```
Comment (ecino) [inline website_compassion/controllers/my_account.py:280]: This is useless computations and filtering.
```suggestion
        contract = request.env["recurring.contract"].browse(int(recurring_contract))
```

File: website_compassion/controllers/my_account.py [hunk 10]
```diff
@@ -269,6 +269,35 @@ def my_child(self, state="active", child_id=None, **kwargs):
...
+        if int(contract.total_amount) > 42:
```
Comment (ecino) [inline website_compassion/controllers/my_account.py:281]:
```suggestion
        if int(contract.total_amount) > 42 or contract.state in ("cancelled", "terminated"):
```

**Review by ecino — CHANGES_REQUESTED** [review]

File: website_compassion/template/my_account_donations.xml [hunk 11]
```diff
@@ -149,7 +150,34 @@
...
+                <t t-foreach="paid_sponsor" t-as="sponsor">
+                    <div class="row">
+                        <div class="col-6">
+                            <label><t t-esc="sponsor.child_id.preferred_name"/></label>
+                        </div>
+                        <div class="col-3">
+                            <label><t t-raw="int(sponsor.total_amount)"/> <t t-esc="sponsor.invoice_line_ids[0].currency_id.name"/></label>
+                        </div>
+                        <div class="col-3" t-if="int(sponsor.total_amount) == 42">
+                          <a t-attf-href="/my/donations/update?recurring_contract=#{sponsor.id}" class="btn btn-primary">Upgrade to 50 <t t-esc="sponsor.invoice_line_ids[0].currency_id.name"/></a>
```
Comment (ecino) [inline website_compassion/template/my_account_donations.xml:174]:
```suggestion
                          <a t-attf-href="/my/donations/upgrade?recurring_contract=#{sponsor.id}" class="btn btn-primary">Upgrade to 50 <t t-esc="sponsor.invoice_line_ids[0].currency_id.name"/></a>
```

File: website_compassion/template/my_account_donations.xml [hunk 12]
```diff
@@ -261,5 +289,17 @@
...
             </div>
         </template>
+        <template id="my_donations_upgrade" name="upgrade">
+            <t t-set="selected">donation</t>
+            <t t-set="title">My financials</t>
+            <t t-call="website_compassion.my_account_layout">
+               <center> Confirm upgrade to 50 CHF ? </center>
+                 <center>
+                    <a t-attf-href="/my/donations/update?res=accepted&amp;recurring_contract=#{sponsor}" class="btn btn-primary">Yes</a>
+                    <a t-attf-href="/my/donations/update?res=refused&amp;recurring_contract=#{sponsor}" class="btn btn-primary">No</a>
```
Comment (ecino) [inline website_compassion/template/my_account_donations.xml:299]:
```suggestion
                    <a t-attf-href="/my/donations/upgrade?res=accepted&amp;recurring_contract=#{sponsor}" class="btn btn-primary">Yes</a>
                    <a t-attf-href="/my/donations/upgrade?res=refused&amp;recurring_contract=#{sponsor}" class="btn btn-primary">No</a>
```

**Pull request closed without merging**